packages/permission_handler/tizen/src/app_settings_manager.cc

```diff
@@ -1,30 +1,116 @@
 
-void AppSettingsManager::OpenAppSettings(OnAppSettingsOpened success_callback,
-                                         OnAppSettingsError error_callback) {
-  app_control_h service = nullptr;
-  int result = app_control_create(&service);
-  if (result != APP_CONTROL_ERROR_NONE) {
-    error_callback(get_error_message(result),
-                   "An error occurred when call app_control_create()");
+  bool Launch(const char* pkg_name) {
+    int ret = app_control_add_extra_data(app_control_, kPackageID, pkg_name);
+    if (ret != APP_CONTROL_ERROR_NONE) {
+      LOG_ERROR("Failed to add key[%s]:value[%s]", kPackageID, pkg_name);
+      return false;
+    }
+
+    ret = app_control_send_launch_request(app_control_, nullptr, nullptr);
+    if (ret != APP_CONTROL_ERROR_NONE) {
+      LOG_ERROR("Failed to send launch request setting");
+      return false;
+    }
+    return true;
   }
 
-  app_control_set_operation(service, APP_CONTROL_OPERATION_SETTING);
-  result = app_control_send_launch_request(service, nullptr, nullptr);
-  app_control_destroy(service);
-  if (result == APP_CONTROL_ERROR_NONE) {
-    LOG_DEBUG("successed to open app settings");
-    success_callback(true);
-  } else {
-    LOG_DEBUG("failed to open app settings");
-    success_callback(false);
+ private:
+  void Init() {
+    int ret = app_control_create(&app_control_);
+    if (ret != APP_CONTROL_ERROR_NONE) {
+      LOG_ERROR("Failed to create app control handle");
+    }
+
+    ret = app_control_set_app_id(app_control_, kSettingAppID);
+    if (ret != APP_CONTROL_ERROR_NONE) {
+      LOG_ERROR("Failed to set app id[%s]", kSettingAppID);
+    }
   }
+
+  void Deinit() {
+    if (app_control_) {
+      app_control_destroy(app_control_);
+      app_control_ = nullptr;
+    }
+  }
+
+  app_control_h app_control_{nullptr};
+};
+
+class PackageName {
+ public:
+  PackageName() { Init(); }
+
+  ~PackageName() { Deinit(); }
+
+  char* Get() { return package_name_; }
+
+ private:
+  void Init() {
+    int ret = app_get_id(&app_id_);
+    if (ret != APP_ERROR_NONE || app_id_ == nullptr) {
+      LOG_ERROR("Failed to get app id");
+      return;
+    }
+
+    ret = package_info_create(app_id_, &package_info_);
+    if (ret != PACKAGE_MANAGER_ERROR_NONE || package_info_ == nullptr) {
+      LOG_ERROR("Failed to create package info handle");
+    }
+    ret = package_info_get_package(package_info_, &package_name_);
+    if (ret != PACKAGE_MANAGER_ERROR_NONE || package_name_ == nullptr) {
+      LOG_ERROR("Failed to get package name");
+    }
+  }
+
+  void Deinit() {
+    if (app_id_) {
+      free(app_id_);
+      app_id_ = nullptr;
+    }
+
+    if (package_info_) {
+      package_info_destroy(package_info_);
+      package_info_ = nullptr;
+    }
+
+    if (package_name_) {
+      free(package_name_);
+      package_name_ = nullptr;
+    }
+  }
+
+  char* app_id_{nullptr};
+  package_info_h package_info_{nullptr};
+  char* package_name_{nullptr};
+};
+}  // namespace
```

Comment:
Is this namespace intended?

Reply:
Yes, I think no need to be exposed outside.